quic/s2n-quic-rustls/src/certificate.rs

```diff
@@ -76,69 +76,85 @@ cert_type!(
-            rustls_pemfile::pkcs8_private_keys,
-        ];
-
-        for parser in parsers.iter() {
-            cursor.set_position(0);
-
-            match parser(&mut cursor) {
-                Ok(keys) if keys.is_empty() => continue,
-                Ok(mut keys) if keys.len() == 1 => {
-                    return Ok(rustls::PrivateKey(keys.pop().unwrap()))
-                }
-                Ok(keys) => {
-                    return Err(Error::General(format!(
-                        "Unexpected number of keys: {} (only 1 supported)",
-                        keys.len()
-                    )));
+        macro_rules! parse_key {
+            ($parser:ident, $key_type:expr) => {
+                cursor.set_position(0);
+
+                let keys: Result<Vec<_>, Error> = rustls_pemfile::$parser(&mut cursor)
+                    .map(|key| {
+                        key.map_err(|_| {
+                            Error::General("Could not load any private keys".to_string())
+                        })
+                    })
+                    .collect();
+                match keys {
+                    // try the next parser
+                    Err(_) => (),
+                    // try the next parser
+                    Ok(keys) if keys.is_empty() => (),
+                    Ok(mut keys) if keys.len() == 1 => {
+                        return Ok($key_type(keys.pop().unwrap()));
+                    }
+                    Ok(keys) => {
+                        return Err(Error::General(format!(
+                            "Unexpected number of keys: {} (only 1 supported)",
+                            keys.len()
+                        )));
+                    }
                 }
-                // try the next parser
-                Err(_) => continue,
-            }
+            };
         }
 
+        // attempt to parse PKCS8 encoded key. Returns early if a key is found
+        parse_key!(pkcs8_private_keys, PrivateKeyDer::Pkcs8);
+        // attempt to parse RSA key. Returns early if a key is found
+        parse_key!(rsa_private_keys, PrivateKeyDer::Pkcs1);
+
         Err(Error::General(
             "could not load any valid private keys".to_string(),
         ))
     }
 }
 
 mod der {
-    use super::*;
+    use rustls::{
+        pki_types::{CertificateDer, PrivateKeyDer},
+        Error,
+    };
 
-    pub fn into_certificate(contents: Vec<u8>) -> Result<Vec<rustls::Certificate>, Error> {
+    pub fn into_certificate(contents: Vec<u8>) -> Result<Vec<CertificateDer<'static>>, Error> {
         // der files only have a single cert
-        Ok(vec![rustls::Certificate(contents)])
+        Ok(vec![CertificateDer::from(contents)])
     }
 
-    pub fn into_private_key(contents: Vec<u8>) -> Result<rustls::PrivateKey, Error> {
-        Ok(rustls::PrivateKey(contents))
+    pub fn into_private_key(contents: Vec<u8>) -> Result<PrivateKeyDer<'static>, Error> {
+        Ok(PrivateKeyDer::Pkcs8(contents.into()))
```

Comment:
rustls now makes use choose that key type when constructing the PrivateKey. Do we need to do something similar to the pem mod and attempt to also parse a Pcks1 and Pksc8 key?

Reply:
Reading up on Pkcs standards: https://stackoverflow.com/a/48960291

Basically PKCS are encoding formats and while PKCS1 can only encode RSA, PKCS8 can encode RSA as well as other key types.

> PKCS1 ... is primarily about using the RSA algorithm for cryptography
> PKCS8 available as [rfc5208](https://www.rfc-editor.org/rfc/rfc5208) on the other hand is a standard for handling private keys for all algorithms, not just RSA.

Reply:
Added a comment in code to avoid future confusion